File: src/patterns/structural/decorator/decorator.ts
interface Pizza {
  toppings: string[];
  price: number;
  getPrice(): number;
  addTopping(...items: string[]): string[];
}

class PizzaBase implements Pizza {
  toppings: string[] = ["bread"];
  price: number = 5;

  constructor(toppings?: string[], price?: number) {
    this.toppings = toppings ? toppings : ["bread"];
    this.price = price ? price : 5;
  }

  getPrice(): number {
    return this.price;
  }

  addTopping(...items: string[]): string[] {
    for (let i = 0; i < items.length; i++) {
      this.toppings.push(items[i]);
    }

    return this.toppings;
  }
}

class BasePizzaDecorator implements Pizza {
  pizzaDecorator: Pizza;
  toppings: string[];
  price: number;

  constructor(pizzaDecorator: Pizza) {
    this.pizzaDecorator = pizzaDecorator;
    this.toppings = pizzaDecorator.toppings;
    this.price = pizzaDecorator.price;
  }

  getPrice(): number {
    return this.pizzaDecorator.getPrice();
  }

  addTopping(...items: string[]): string[] {
    return this.pizzaDecorator.addTopping(...items);
  }
}

class PizzaWithCheese extends BasePizzaDecorator {
  constructor(pizzaDecorator: Pizza) {
    super(pizzaDecorator);
  }

  getPrice(): number {
    return this.pizzaDecorator.getPrice() + 5;
  }

  addTopping(...items: string[]): string[] {
    return this.pizzaDecorator.addTopping(...items, "cheese");
  }
}

class PizzaWithPepperoni extends BasePizzaDecorator {
  constructor(pizzaDecorator: Pizza) {
    super(pizzaDecorator);
  }

  getPrice(): number {
    return this.pizzaDecorator.getPrice() + 10;
  }

  addTopping(...items: string[]): string[] {
    return this.pizzaDecorator.addTopping(...items, "pepperoni");
  }
}

let margharitta = new PizzaBase();
margharitta = new PizzaWithCheese(margharitta);
console.log(margharitta.getPrice());
let pepperoni = new PizzaWithPepperoni(new PizzaWithCheese(new PizzaBase()));
console.log(pepperoni.getPrice());
console.log(pepperoni.addTopping("pepperoni"));

File: src/patterns/structural/decorator/decorator.go
package main

import (
	"fmt"
	"strings"
)

// Component
type PizzaInterface interface {
	GetPrice() int
	GetDescription() string
}

// Concrete Component

type Pizza struct {	
	topping []string;
	price int
}

func (p *Pizza) GetPrice() int {
	return p.price
}

func (p *Pizza) GetDescription() string {
	var sb strings.Builder
	initialString := "The toppings on the pizza include: "
	sb.WriteString(initialString)
	for i := 0; i < len(p.topping); i++ {
		sb.WriteString(p.topping[i])
	}
	return sb.String()
}

// Base Decorator

type PizzaDecorator struct {
	decoratedPizza PizzaInterface
}

func (p *PizzaDecorator) GetPrice() int {
	return p.decoratedPizza.GetPrice()
}

func (p *PizzaDecorator) GetDescription() string {
	return p.decoratedPizza.GetDescription()
}

// Concrete Decorators

type WithCheese struct {
	decoratedPizza PizzaInterface
}

func (p *WithCheese) GetPrice() int {
	return p.decoratedPizza.GetPrice() + 10
}

func (p *WithCheese) GetDescription() string {
	return p.decoratedPizza.GetDescription() + " cheese"
}


type WithMeat struct {
	decoratedPizza PizzaInterface
}

func (p *WithMeat) GetPrice() int {
	return p.decoratedPizza.GetPrice() + 15
}

func (p *WithMeat) GetDescription() string {
	return p.decoratedPizza.GetDescription() + " meat"
}

// Usage

func main() {
	basePizza := &Pizza{topping: []string{"bread"}, price: 100}
	pizzaWithCheese := &WithCheese{decoratedPizza: basePizza}
	fmt.Println(pizzaWithCheese.GetPrice())
	fmt.Println(pizzaWithCheese.GetDescription())
	pizzaWithMeat := &WithMeat{decoratedPizza: pizzaWithCheese}
	fmt.Println(pizzaWithMeat.GetPrice())
	fmt.Println(pizzaWithMeat.GetDescription())
}
